### ISSUE TITLE
[question] How to prevent `conan install` from polluting the source repository?

### ISSUE THREAD

**Issue by PauloCarvalhoRJ** [conan-io/conan#17026]
Greetings,

So, I have a CI/CD script that builds packages with `conan install` commands (e.g.  `conan install ninja/1.12.1 -s build_type=Release --build=missing --build=outdated --build=ninja`).  The commands complete but they put the build artifacts in the source directory instead of in the `b` directory in Conan cache like the `conan create` does.  How do I make `conan install` behave similarly to `conan create`?

thank you,

PC

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @PauloCarvalhoRJ 

Which Conan version? ``--build=outdated`` only existed in Conan 1, in Conan 2 it is useless. Furthermore the syntax ``conan install pkg/version`` is not correct neither in Conan 1 nor in Conan 2.

Could you please clarify the exact Conan version, the exact command that you are using, the output of the command, report what is inside the respective folders (which folders they are too) in the cache and compare that with the folders with ``conan create``?

**Comment by PauloCarvalhoRJ** [comment]
Hello.  Sorry , it's Conan 2.3.

**Comment by memsharded** [comment]
> Hello. Sorry , it's Conan 2.3.

And what would be the correct command then? and the output, and the resulting folder contents?

**Comment by PauloCarvalhoRJ** [comment]
Ok.

The script forms the following command for Ninja: `conan install ninja/1.12.1 -s build_type=Release --build=missing --build=outdated --build=ninja`.  The Ninja recipe is rather simple and only extracts the correct executable from a zip file to the package directory.

So, this command results in six shell scripts copied to the source diretcory:
![image](https://github.com/user-attachments/assets/61e3e7a6-a1fa-4f0a-ab72-1e04ccef0abf)

If I run `conan create ninja/1.12.1`, however, I get everything where they are supposed to be:
```bash
/home/conan2_packages/
|-- b
|   `-- ninja73b1cb0514cd0
|       |-- b
|       |   |-- conanbuildenv.sh
|       |   |-- conanbuild.sh
|       |   |-- conaninfo.txt
|       |   |-- conanrunenv.sh
|       |   |-- conanrun.sh
|       |   |-- deactivate_conanbuild.sh
|       |   |-- deactivate_conanrun.sh
|       |   |-- ninja
|       |   `-- ninja.exe
|       |-- d
|       |   `-- metadata
|       `-- p
|           |-- bin
|           |   `-- ninja
|           |-- conaninfo.txt
|           `-- conanmanifest.txt
|-- cache.sqlite3
|-- ninja20bbd65455361
|   |-- d
|   |   `-- metadata
|   |-- e
|   |   |-- conanfile.py
|   |   `-- conanmanifest.txt
|   |-- es
|   |   `-- ninja.zip
|   `-- s
|       `-- ninja.zip
`-- t
```

**Comment by PauloCarvalhoRJ** [comment]
Notice that the six scripts from the `conan install ...` are in cache.

**Comment by memsharded** [comment]
The commands ``conan install ninja/1.12.1`` and ``conan create ninja/1.12.1`` are incorrect in Conan 2.

**Comment by PauloCarvalhoRJ** [comment]
But `conan create ninja/1.12.1` is working as expected.

**Comment by memsharded** [comment]
```
(conan2_310) λ conan create ninja/1.2.11
ERROR: Conanfile not found at C:\Users\memsharded\conanws\kk\ninja\1.2.11
```
this only works if you have a folder called ``ninja`` with a subfolder called ``1.2.11``, but that is not the case

**Comment by PauloCarvalhoRJ** [comment]
Ah... yes.  My recipes are customized to have a fixed version. `1.2.11` in my case is the directory where the recipe is.   Think `conan create ninja/1.12.1` as `conan create ninja` or even `conan create .` when running from the same directory where the recipe is.

**Comment by PauloCarvalhoRJ** [comment]
Recall that I'm on Linux, so `conan create ninja/1.12.1` should read `conan create ninja\1.12.1` in Windows.

**Comment by PauloCarvalhoRJ** [comment]
Would you like to share my recipe/zip?  They're quite small.

**Comment by memsharded** [comment]
> Would you like to share my recipe/zip? They're quite small.

yes, please, if you are using a custom recipe, and not a known one from ConanCenter we need it to be able to reproduce.

**Comment by PauloCarvalhoRJ** [comment]
Here:

[ninja.zip](https://github.com/user-attachments/files/17075941/ninja.zip)

[make_life_simple_please.zip](https://github.com/user-attachments/files/17075982/make_life_simple_please.zip)

On a side note: why doesn't github accept certain file extensions?  It only complicates life unnecessarily...

The recipe and the profile I'm using (for both host and build contexts) are inside the other zip.

Thank you!

**Comment by valgur** [comment]
Your conanfile.py is probably missing a `layout()` section with either a `cmake_layout(self)`, `basic_layout(self)` or similar as contents. Without one you need to pass `-of` to the install command to have the generated build files be placed in some subdir.

I have not checked in a while, but I think the Conan tutorial does not emphasize this detail all that well.

**Comment by PauloCarvalhoRJ** [comment]
Hello, @valgur ,

That Ninja recipe is very simple and does not have a `layout()` callback.  So far, I've never had the need for it.  Anyway, I'm going to try it and report here the result.

thanks,

PC

**Comment by PauloCarvalhoRJ** [comment]
So, @valgur , adding the following callback:
```python
    def layout(self):
        basic_layout(self, src_folder="src")
```
improved the result of the `conan install ...` in a sense.  The six forementioned Conan shell scripts are now in a `build` folder in the source repository and it is not "building".  That is, it is not unzipping the Ninja executable to `<Conan cache path>/b/ninja<hash>/p/bin/ninja`.  So, there is still a long way to go.

thanks,

PC

**Comment by memsharded** [comment]
Thanks very much for your feedback and code to help reproduce @PauloCarvalhoRJ, that really helps.

Indeed, as @valgur suggests, the issue is a lack of ``layout()`` method. This method is the one that allows to have a consistent layout of folders both in the local user space (when running ``conan install``) and when creating in the Conan cache.

If you don't need the environment files for your recipe, it is likely that the most straightforward approach is to do:
```python
def generate(self):
    VirtualBuildEnv(self)
    VirtualRunEnv(self)
```
which disables the creation of the files completely.

**Comment by PauloCarvalhoRJ** [comment]
Hello, @memsharded ,

Sorry for the late response.  Well, that was yet another small improvement, nevertheless there are two issues pending:
1) An empty `build/conan` path is still being created in the source repository.
2) More importantly, nothing is being built (only a file named `cache.sqlite3` is being created in Conan cache).

Recall the command line: `conan install ninja/1.12.1 -s build_type=Release --build=missing --build=outdated --build=ninja` .

thank you,

PC

**Comment by memsharded** [comment]
> More importantly, nothing is being built (only a file named cache.sqlite3 is being created in Conan cache).

That is expected. ``conan install <path/to/ninja/recipe>`` will install the dependencies of the ninja recipe, build them if necessary, but this recipe doesn't have dependencies, so nothing is built.

**Comment by PauloCarvalhoRJ** [comment]
Than what are the command line arguments `-s build_type=Release --build=missing --build=outdated --build=ninja` supposed to mean?

**Comment by memsharded** [comment]
They will be building missing and outdated dependencies, and also dependencies matching the ``ninja`` name. As there are no dependencies of the current recipe, there are no dependencies that matches those rules either, so nothing is built. Note that ``ninja`` is not a dependency of itself, the ``conan install <path/to/ninja/recipe>`` means install the dependencies of that, but ninja is not a dependency to itself.

**Comment by PauloCarvalhoRJ** [comment]
So, how do I get rid of the `build` directory that is still polluting the source repository?

**Comment by memsharded** [comment]
If you don't want ``conan install`` to install any generator, and you don't want the build folder created you can define:
```python
def layout(self):
        self.folders.build = "."
```
That means that basically ``conan install`` is a no-op, because there are no dependencies, and there are no generators either, not for dependencies, not for toolchains, nothing. So ``conan install`` does nothing, you might just avoid calling it.

**Comment by PauloCarvalhoRJ** [comment]
> > More importantly, nothing is being built (only a file named cache.sqlite3 is being created in Conan cache).
> 
> That is expected. `conan install <path/to/ninja/recipe>` will install the dependencies of the ninja recipe, build them if necessary, but this recipe doesn't have dependencies, so nothing is built.

So, I tried to `conan install` a consumer package: `conan install cmake/3.29.3 -s build_type=Release --build=missing --build=outdated` and I got this error:
```bash
ninja/1.12.1: Not found in local cache, looking in remotes...
Graph root
    conanfile.py (cmake_installer/3.29.3): /Work3/ur5m/GammaRay/gammaray_src/conan2_recipes/cmake/3.29.3/conanfile.py
ERROR: Package 'ninja/1.12.1' not resolved: No remote defined.
```
So, the dependency (Ninja in this case) is not being built on demand following your rationale.

The CMake recipe I'm using (along with the CMake binaries) can be found here: https://github.com/PauloCarvalhoRJ/gammaray/tree/build_with_conan/conan2_recipes/cmake/3.29.3 .

best,

PC

**Comment by PauloCarvalhoRJ** [comment]
> If you don't want `conan install` to install any generator, and you don't want the build folder

Sorry if I'm still not clear.  That was never the issue.  The issue is that I need them in the Conan cache like `conan create` does and not in the source repository.

**Comment by memsharded** [comment]
The problem is not that ``ninja`` is not being built, it is not a "missing binary" error.
It is just not finding the ``ninja/1.12.1`` anywhere.

The cmake recipe contains a `` requires = "ninja/1.12.1"`` which is by itself probably a bad practice, it should be in the best case a ``tool_requires``, as ``requires`` are for libraries dependencies not tools.

So when you are ``conan install <path/to/cmake/recipe>`` the ``ninja/1.12.1`` package must be already existing, either in the cache or in a server, but in the above error message it isn't. Conan cannot invent or build the ``ninja/1.12.1`` package out of nowhere, it needs the package to be there first in order to use it.

**Comment by memsharded** [comment]
> Sorry if I'm still not clear. That was never the issue. The issue is that I need them in the Conan cache like conan create does and not in the source repository.

Then, we might need to re-formulate this ticket, ``conan create`` works in the cache, with different behavior than in the local dev-flow in user space.

**Comment by PauloCarvalhoRJ** [comment]
Returning to the original question: How do I do a `conan install` without anything being output to the source repository?

**Comment by memsharded** [comment]
You can use:
- ``--output-folder``: like ``conan install . --output-folder=../mybuild`` will create a ``mybuild`` folder, as a sibling of the current one. This will have internally a ``build/...`` root lib
- ``conan install . -c tools.cmake.cmake_layout:build_folder=../mybuild``. This will not have an internal ``build`` subfolder.

In any case, this approach is not practical. The recommended approach for most cases is to use layouts such as the ``cmake_layout()`` and ``.gitignore`` the "build" folder.

**Comment by PauloCarvalhoRJ** [comment]
That used to work in Conan 1 without polluting the source directory nor requiring complicated command line arguments.

**Comment by memsharded** [comment]
Not really, Conan 1 creates also the ``build`` folder with the generators there, and also a bunch of other files like ``conaninfo.txt``, ``conanbuildinfo.txt``, ``graph_info.json``, which cannot be skipped in any way. So Conan 1 was polluting more than Conan 2. It is easy to replicate:
``` bash
# with Conan 1.65
conan new hello/0.1 -m cmake_lib
conan install .
```
And you will see the extra files there.

**Comment by PauloCarvalhoRJ** [comment]
Anyway, how do I get Ninja build files to the Conan cache?  By adding a `layout()` callback, I got the build files in a `build` directory in the source repository.  I can live with that (requires adding this to the `.gitignore` file).

However, if nothing is being copied to the Conan cache, when I run `conan install` for a consumer package, I get that `ninja/1.12.1: Not found in local cache` error.

So, I still need something copied/generated in the Conan cache like the recipe used to do without the `layout()` callback.

In a nutshell: How do I keep the build artifacts in the `build` directory AND generate the required files in the Conan cache so consumer packages work?

Thank you

**Comment by PauloCarvalhoRJ** [comment]
> So, I still need something copied/generated in the Conan cache like the recipe used to do without the `layout()` callback.

I got this by doing a prior `conan export`.  Sorry for forgetting this.

**Comment by memsharded** [comment]
The process for getting something to become a package is to put it in the cache. This can be done with:

- ``conan export`` It puts only the recipe part, ``conanfile.py`` and associated files, but no binaries, so the binaries will always be missing and need to built from source.
- ``conan create``, does the whole process ``conan export .`` to put the recipe in the cache, ``conan install --requires=<pkg/version> --build=<pkg/version>`` to force the build of the binary in the cache, then launches the ``conan test test_package`` if it exists
- ``conan export-pkg`` does the ``conan export .`` to put the recipe, the proceeds to do a direct export of the binaries from the local user folder to the cache (package pre-compiled binaries from user folders), using the ``package()`` method. Then it will also call ``conan test test_package`` in Conan 2 if it exists.

The create and export-pkg will end with package binaries in the cache if done correctly, ready to be used by other consumers.

**Comment by PauloCarvalhoRJ** [comment]
That was a nice summary, thanks.  Now the script to build all packages in a single shot respecting their dependency order is working.  Though, there will be `build` directories in the source repository.  Maybe their removal can be automated or a simple addition to the `.gitignore` file will suffice.
